content/blog/2022-03-08-alive/index.md

```diff
@@ -0,0 +1,114 @@
++++
+title = "Efficient Path Profiling"
+[[extra.authors]]
+name = "Aaron Tucker"
+link = "https://www.cs.cornell.edu/~aarondtucker/"
++++
+
+## Blog: Provably Correct Peephole Optimizations with Alive
+
+Citation: Nuno P. Lopes, David Menendez, Santosh Nagarakatte, and John Regehr. 2015. Provably correct peephole optimizations with alive. SIGPLAN Not. 50, 6 (June 2015), 22–32. DOI:https://doi.org/10.1145/2813885.2737965
+
+## Introduction
+Provably Correct Peephole Optimizations with Alive by Nuno P. Lopes, David Menendez, Santosh Nagarakatte, and John Regehr (SIGPLAN 2015) is a paper with a pretty straightforward goal -- define a domain specific language and semantics for peephole optimizations in LLVM which make it easy to verify the correctness of peephole optimizations, while also building tools which make it simple to actually use the domain specific language when _you_ write peephole optimizations. Overall, I think this paper is great at scoping the project to be tractable and impactful, and is also rather elegant in how it leverages various properties of SMT solvers to make a better tool. Before we get into all of that, there is some preliminary background.
+
+
+## Background
+At its core, Alive is a domain specific language for formally verifying peephole optimizations in LLVM. Empirically, it found bugs where the optimization could introduce additional undefined behavior in situations where the original code did not. For me, the main important pieces of background to understand were LLVM, Peephole Optimizations, SMT Solvers, and Undefined Behaviors in C/C++.
+
+### LLVM
+According to [the project site](https://llvm.org), LLVM is “a collection of modular and reusable compiler and toolchain technologies”. Basically, compilers have a lot of similar components – they need to split up the program’s control flow to define all the basic blocks, perform dead code elimination, rearrange loops, and many other optimizations. LLVM makes it easier to reuse and compose these components by creating an _intermediate representation_ which is much closer to being machine code, but still has enough complexity and expressiveness to be able to figure out and perform the optimizations on a representation which is not yet specific to a particular architecture, and which is easier to think about.
+
+This project has an active development community and people do contribute individual optimizations and pieces of code, so architecturally this seems to have been a great choice.
```

Comment:
It may also be relevant (w/r/t why to choose LLVM as a verification target) that LLVM is extremely widely _used_, so any bug in it will potentially affect a lot of compiled code.